magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

KrasnoshchokBohdan opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

**First part -**
In this issue, the author mentioned that responce from 
<details>
  <summary>this mutation</summary>

```graphql
mutation {
  addProductsToCart(
    cartId: "LclKxxxxxxxxxxxxxxW"
    cartItems: [
      {
        quantity: 1
        parent_sku: "WSH12"
        sku: "WSH12-28-Green"
      }
    ]
  ) {
    cart {
      itemsV2 {
        items {
          id
          product {
            name
            sku
          }
          quantity
        }
        total_count
        page_info {
          page_size
          current_page
          total_pages
        }
      }
    }
    user_errors {
      code
      message
    }
  }
}
```
</details>
doesn't match 
<details>
  <summary>this expected result</summary>

```graphql
{
  "data": {
    "addProductsToCart": {
      "cart": {
        "itemsV2": {
          "items": [
            {
              "id": "24",
              "product": {
                "name": "Erika Running Short",
                "sku": "WSH12"
              },
              "quantity": 1
            },
            {
              "id": "26",
              "product": {
                "name": "Erika Running Short-28-Green",
                "sku": "WSH12-28-Green"
              },
              "quantity": 1
            }
          ],
          "total_count": 2,
          "page_info": {
            "page_size": 20,
            "current_page": 1,
            "total_pages": 1
          }
        }
      },
      "user_errors": []
    }
  }
}
```
</details>
instead, we get 
<details>
  <summary>this actual result</summary>

```graphql
{
  "data": {
    "addProductsToCart": {
      "cart": {
        "itemsV2": {
          "items": [
            {
              "id": "18",
              "product": {
                "name": "Erika Running Short-28-Green",
                "sku": "WSH12-28-Green"
              },
              "quantity": 1
            }
          ],
          "total_count": 1,
          "page_info": {
            "page_size": 20,
            "current_page": 1,
            "total_pages": 1
          }
        }
      },
      "user_errors": []
    }
  }
}
```
</details>

let's add a configurable product to the cart from frontend
<img width="1336" alt="image" src="https://github.com/user-attachments/assets/12245b9f-5601-4409-a30e-b357c1c6eac4" />
in the "quote_item" table we have 2 items
<img width="620" alt="image" src="https://github.com/user-attachments/assets/eed63a72-f525-43ed-b2f5-240fece03564" />

let's try this query 
<img width="1238" alt="image" src="https://github.com/user-attachments/assets/7fa118c5-afdd-459a-ab24-160041514c41" />
and in the responce, we see only the configurable product
the reason for this is 
\Magento\QuoteGraphQl\Model\Resolver\CartItemsPaginated::resolve
\Magento\Quote\Model\Quote::getAllVisibleItems
<img width="1007" alt="image" src="https://github.com/user-attachments/assets/67e9dbc7-a91f-44c9-b564-2a89f521aed2" />
(item with sku "test simple" has parent_item_id)

so maybe it's better to change the responce example on [this](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/add-products/#specify-a-parent-and-a-child-sku) page instead of changing  the code in \Magento\Quote\Model\Quote::getAllVisibleItems 

**Second part -**
let's add a configurable product to the cart using the "addProductsToCart" mutation

<img width="1337" alt="image" src="https://github.com/user-attachments/assets/8ac1e50f-e567-4926-b65c-510867285c65" />

in the responce we see only "test simple", and in "quote_item" table we have only one item 
<img width="617" alt="image" src="https://github.com/user-attachments/assets/f82b238d-c593-4011-b0f0-8a4681295a30" />

so I created ConfigurableProductPrecursor to add a configurable product to the cart along with a simple product.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#[39873](https://github.com/magento/magento2/issues/39873)

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1.create an empty cart [the official doc](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/create-guest-cart/)
2.add a product by specifying parent_sku following [the official doc](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/add-products/#specify-a-parent-and-a-child-sku)


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

KrasnoshchokBohdan committed:
magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

- adding new ConfigurableProductPrecursor

KrasnoshchokBohdan committed:
magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

- modifying ConfigurableProductPrecursor

KrasnoshchokBohdan committed:
Update AddProductsToCart.php

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39873

KrasnoshchokBohdan committed:
magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

- Introduced comprehensive unit tests for ConfigurableProductPrecursor, covering scenarios like valid configurations, non-configurable parents, missing attributes, and missing products. Minor code adjustments were also made for improved readability and array merging in the process method.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39873

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39873

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39873

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/CartItem/Precursor/ConfigurableProductPrecursor.php:
```diff
@@ -0,0 +1,147 @@
+     * @param ProductRepositoryInterface $productRepository
+     */
+    public function __construct(
+        ProductRepositoryInterface $productRepository
```
User `readonly` here

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the collaboration!

Please look into the failed automated test, it seems they are failing due to the code changes. 

Thanks

KrasnoshchokBohdan committed:
magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

Refactor constructor and update error message formatting

Replaced manual property assignment in the constructor with a `readonly` property for `ProductRepositoryInterface` to simplify initialization. Improved code readability by adjusting the formatting of the error message array.

KrasnoshchokBohdan commented:
@engcom-Hotel Hi! I stuck with WebAPI tests, in report I see /var/www/html/dev/tests/api-functional/phpunit.xml
but there is no such file in the magento2 repository 
<img width="1500" height="668" alt="image" src="https://github.com/user-attachments/assets/966a45e6-0f31-4c84-854c-2228732f5fbf" />

and all my attempts with dev/tests/api-functional/phpunit_graphql.xml leads to errors 
https://developer.adobe.com/commerce/webapi/graphql/develop/functional-testing/

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39873: addProductsToCart mutation doesn't work as expect with a given parent_sku

Refactor unused parameter in ConfigurableProductPrecursor.

Renamed the parameter to align with coding standards and added a suppression for PHPMD warnings about unused parameters. This change improves code clarity and adheres to best practices.

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Hotel commented:
> @engcom-Hotel Hi! I stuck with WebAPI tests, in report I see /var/www/html/dev/tests/api-functional/phpunit.xml
> but there is no such file in the magento2 repository

Hello @KrasnoshchokBohdan,

By default, the file is named with the `.dist` suffix and is located at `dev/tests/api-functional/`. To proceed, please rename it to `phpunit.xml` in the same directory. For further details on running the tests, you may refer to the following documentation:

https://developer.adobe.com/commerce/webapi/get-started/web-api-functional-testing/#running-the-tests

I hope this addresses your question.

Thanks

engcom-Dash committed:
Merge branch '2.4-develop' into fix-for-issue-39873

engcom-Dash commented:
Hello @KrasnoshchokBohdan,

We have sent an Email to the Product Owner on the clarification of the expected behaviour for both mutation & query. Till then we are moving this PR on Hold.

Thanks!

engcom-Dash commented:
Hello @KrasnoshchokBohdan,

Yes, as per the Product Owner’s feedback from Olena [comment](https://github.com/magento/magento2/pull/39950#issuecomment-3188849012)  the expected behavior for both the mutation and query is to display the child products. This is consistent with what we already observe in the 2.4-develop branch.

I have verified this in the 2.4-develop environment
<img width="3420" height="1688" alt="image" src="https://github.com/user-attachments/assets/7feafca8-4723-44f2-b53e-612a687f82d7" />


Additionally, the behavior aligns with the documentation reference mentioned in the [Add configurable product to cart mutation](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/add-products/#add-a-configurable-product-to-a-cart).

So, as the expected behavior matches the current implementation in 2.4-develop and aligns with the Product Owner’s confirmation, we are good to close this PR. 

If you have any additional feedback or suggestions regarding this PR, please feel free to reopen it or create a separate PR.

Thank you for your contribution!
